basic_pitch/data/tf_example_deserialization.py

```diff
@@ -0,0 +1,697 @@
+    for ds in [ds_train, ds_validation]:
+        tf.debugging.assert_none_equal(
+            tf.cast(tf.data.experimental.cardinality(ds), tf.int32),
+            tf.data.experimental.INFINITE_CARDINALITY,
+        )
+
+    # training dataset
+    if training_shuffle_buffer_size > 0:
+        # Lets try to cache before the shuffle. This is the entire training dataset so we'll cache
+        # to memory
+        ds_train = (
+            ds_train.shuffle(training_shuffle_buffer_size, reshuffle_each_iteration=True)
+            .repeat()
+            .batch(batch_size)
+            .prefetch(tf.data.AUTOTUNE)
+        )
+
+    # validation dataset
+    ds_validation = (
+        ds_validation.repeat()
+        .batch(batch_size)
+        .take(validation_steps)
+        .cache(f"validation_set_cache_{str(uuid.uuid4())}")
+        .repeat()
+        .prefetch(tf.data.AUTOTUNE)
+    )
+
+    return ds_train, ds_validation
+
+
+def prepare_visualization_datasets(
+    datasets_base_path: str,
+    batch_size: int,
+    validation_steps: int,
+    datasets_to_use: List[str],
+    dataset_sampling_frequency: np.ndarray,
+) -> Tuple[tf.data.Dataset, tf.data.Dataset]:
+    """
+    Return a training and a testing dataset.
+
+    Args:
+        training_shuffle_buffer_size : size of shuffle buffer (only for training set)
+        batch_size : ..
+    """
+
+    assert batch_size > 0
+    assert validation_steps is not None and validation_steps > 0
+
+    ds_train = sample_datasets(
+        "train",
+        datasets_base_path,
+        datasets=datasets_to_use,
+        dataset_sampling_frequency=dataset_sampling_frequency,
+        n_samples_per_track=1,
+    )
+    ds_validation = sample_datasets(
+        "validation",
+        datasets_base_path,
+        datasets=datasets_to_use,
+        dataset_sampling_frequency=dataset_sampling_frequency,
+        n_samples_per_track=1,
+    )
+
+    # check that the base dataset returned by ds_function is FINITE
+    for ds in [ds_train, ds_validation]:
+        tf.debugging.assert_none_equal(
+            tf.cast(tf.data.experimental.cardinality(ds), tf.int32),
+            tf.data.experimental.INFINITE_CARDINALITY,
+        )
+
+    # training dataset
+    ds_train = ds_train.repeat().batch(batch_size).prefetch(tf.data.AUTOTUNE)
+
+    # validation dataset
+    ds_validation = (
+        ds_validation.repeat()
+        .batch(batch_size)
+        .take(validation_steps)
+        .cache(f"validation_set_cache_{str(uuid.uuid4())}")
+        .repeat()
+        .prefetch(tf.data.AUTOTUNE)
+    )
+
+    return ds_train, ds_validation
+
+
+def sample_datasets(
+    split: str,
+    datasets_base_path: str,
+    datasets: List[str],
+    dataset_sampling_frequency: np.ndarray,
+    n_shuffle: int = 1000,
+    n_samples_per_track: int = N_SAMPLES_PER_TRACK,
+    pairs: bool = False,
+) -> tf.data.Dataset:
+    assert split in ["train", "validation"]
```

Comment:
Extra credit: Can we turn the split str into an Enum?

Reply:
would we want to do this throughout he repo?